File: Day 18- Binary Search/Maximize the Minimum Importance.java
class Solution {
    public static int maxMinImportance(int n, int r, int k, int[] arr) {
        long left = 0, right = k;
        for (int x : arr) right += x;
        long[] v = new long[n];
        while (left <= right) {
            long x = (left + right) / 2;
            for (int i = 0; i < n; ++i) v[i] = arr[i];
            long s = 0, use = 0;
            for (int i = 0; i < r; ++i) s += v[i];
            for (int i = 0; i < n; ++i) {
                // add to t if needed
                int t = Math.min(n - 1, i + r);
                // update sd
                if (i + r < n) s += v[i + r];
                if (i - r > 0) s -= v[i - r - 1];
                long diff = Math.max(0, x - s);
                v[t] += diff;
                s += diff;
                use += diff;
            }
            if (use <= k)
                left = x + 1;
            else
                right = x - 1;
        }
        return (int)right;
    }
}

File: Day 18, 19 - Binary Search/Maximize the Minimum Importance.cpp
// Back-end complete function Template for C++

class Solution
{
public:
  bool isGood(vector<int> &arr, int r, long long minPowerRequired, int additionalarr)
  {
    int n = arr.size();
    // init windowPower to store power of 0th city (minus arr[r])
    long long windowPower = accumulate(arr.begin(), arr.begin() + r, 0LL);
    vector<int> additions(n, 0);
    for (int i = 0; i < n; i++)
    {
      if (i + r < n)
      {
        // now, windowPower stores sum of power arr from [i-r..i+r],
        // it also means it's the power of city `ith`
        windowPower += arr[i + r];
      }
      if (windowPower < minPowerRequired)
      {
        long long needed = minPowerRequired - windowPower;
        if (needed > additionalarr)
        {
          // Not enough additional arr to plant
          return false;
        }
        // Plant the additional arr on the farthest city in the range
        // to cover as many cities as possible
        additions[min(n - 1, i + r)] += needed;
        windowPower = minPowerRequired;
        additionalarr -= needed;
      }
      if (i - r >= 0)
      {
        // out of window range
        windowPower -= arr[i - r] + additions[i - r];
      }
    }
    return true;
  }

  long long maxMinImportance(int n, int r, int k, vector<int> &arr)
  {
    long long left = 0;
    // The answer = `right`, when `r = n`, all value of arr are the same!
    long long right = accumulate(arr.begin(), arr.end(), 0LL) + k;
    long long ans = 0;
    while (left <= right)
    {
      long long mid = (left + right) / 2;
      if (isGood(arr, r, mid, k))
      {
        ans = mid;      // This is the maximum possible minimum power so far
        left = mid + 1; // Search for a larger value in the right side
      }
      else
      {
        right = mid - 1; // Decrease minPowerRequired to need fewer additional power arr
      }
    }
    return ans;
  }
};
